dotnet/src/Connectors/Connectors.HuggingFace/Core/HuggingFaceClient.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.SemanticKernel.Connectors.HuggingFace.TextGeneration;
+using Microsoft.SemanticKernel.Http;
+
+namespace Microsoft.SemanticKernel.Connectors.HuggingFace.Core;
+
+internal sealed class HuggingFaceClient : IHuggingFaceClient
+{
+    private readonly IStreamJsonParser _streamJsonParser;
+    private readonly string _modelId;
+    private readonly string? _apiKey;
+
+    private IHttpRequestFactory HttpRequestFactory { get; }
+    private IEndpointProvider EndpointProvider { get; }
+    private HttpClient HttpClient { get; }
+    private ILogger Logger { get; }
+
+    internal HuggingFaceClient(
+        string modelId,
+        HttpClient httpClient,
+        IHttpRequestFactory httpRequestFactory,
+        IEndpointProvider endpointProvider,
+        string? apiKey = null,
+        IStreamJsonParser? streamJsonParser = null,
+        ILogger? logger = null)
+    {
+        Verify.NotNullOrWhiteSpace(modelId);
+
+        this._modelId = modelId;
+        this._apiKey = apiKey;
+        this.HttpClient = httpClient;
+        this.HttpRequestFactory = httpRequestFactory;
+        this.EndpointProvider = endpointProvider;
+        this.Logger = logger ?? NullLogger.Instance;
+        this._streamJsonParser = streamJsonParser ?? new TextGenerationStreamJsonParser();
+    }
+
+    public async Task<IReadOnlyList<TextContent>> GenerateTextAsync(string prompt, PromptExecutionSettings? executionSettings, CancellationToken cancellationToken)
+    {
+        var endpoint = this.EndpointProvider.GetTextGenerationEndpoint(executionSettings?.ModelId ?? this._modelId);
+        var request = this.CreateTextRequest(prompt, executionSettings);
+        using var httpRequestMessage = this.HttpRequestFactory.CreatePost(request, endpoint, this._apiKey);
+
+        string body = await this.SendRequestAndGetStringBodyAsync(httpRequestMessage, cancellationToken)
+            .ConfigureAwait(false);
+
+        var response = DeserializeResponse<TextGenerationResponse>(body);
+        var textContents = GetTextContentFromResponse(response, executionSettings?.ModelId ?? this._modelId);
+
+        this.LogTextGenerationUsage(executionSettings);
+
+        return textContents;
+    }
+
+    public async IAsyncEnumerable<StreamingTextContent> StreamGenerateTextAsync(
+        string prompt,
+        PromptExecutionSettings? executionSettings = null,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
```

Comment:
Considering that this class is an internal one, I would suggest always requiring the cancellation token argument to avoid forgetting to pass it from the calling method. Same for the GenerateEmbeddingsAsync method.